example/plugins/c-api/txn_data_sink/txn_data_sink.cc

```diff
@@ -28,89 +28,91 @@
   limitations under the License.
  */
 
-#include <stdio.h>
-#include <unistd.h>
 #include <ts/ts.h>
 
-// This gets the PRI*64 types
-#define __STDC_FORMAT_MACROS 1
-#include <inttypes.h>
+#include <string>
+#include <string_view>
 
-#define PLUGIN_NAME "txn_data_sink"
-#define PCP "[" PLUGIN_NAME "] "
+namespace
+{
+std::string const PLUGIN_NAME = "txn_data_sink";
 
-// Activate the data sink if this field is present in the request.
-static const char FLAG_MIME_FIELD[] = "TS-Agent";
-static size_t const FLAG_MIME_LEN   = sizeof(FLAG_MIME_FIELD) - 1;
+/** Activate the data sink if this header field is present in the request. */
+std::string const FLAG_HEADER_FIELD = "TS-Agent";
 
-typedef struct {
-  int64_t total;
-} SinkData;
+/** The sink data for a transaction. */
+struct SinkData {
+  /** The bytes for the response body streamed in from the sink. */
+  std::string body_bytes;
+};
 
 // This serves to consume all the data that arrives. If it's not consumed the tunnel gets stalled
 // and the transaction doesn't complete. Other things could be done with the data, accessible via
 // the IO buffer @a reader, such as writing it to disk to make an externally accessible copy.
-static int
+int
 client_reader(TSCont contp, TSEvent event, void *edata)
 {
-  SinkData *data = TSContDataGet(contp);
+  SinkData *data = static_cast<SinkData *>(TSContDataGet(contp));
 
   // If we got closed, we're done.
   if (TSVConnClosedGet(contp)) {
-    TSfree(data);
+    delete data;
     TSContDestroy(contp);
     return 0;
   }
 
   TSVIO input_vio = TSVConnWriteVIOGet(contp);
 
   if (!data) {
-    data        = TSmalloc(sizeof(SinkData));
-    data->total = 0;
+    data = new SinkData;
     TSContDataSet(contp, data);
   }
 
   switch (event) {
   case TS_EVENT_ERROR:
-    TSDebug(PLUGIN_NAME, "Error event");
+    TSDebug(PLUGIN_NAME.c_str(), "Error event");
     TSContCall(TSVIOContGet(input_vio), TS_EVENT_ERROR, input_vio);
     break;
   case TS_EVENT_VCONN_READ_COMPLETE:
-    TSDebug(PLUGIN_NAME, "READ_COMPLETE");
+    TSDebug(PLUGIN_NAME.c_str(), "READ_COMPLETE");
     break;
   case TS_EVENT_VCONN_READ_READY:
   case TS_EVENT_IMMEDIATE:
-    TSDebug(PLUGIN_NAME, "Data event - %s", event == TS_EVENT_IMMEDIATE ? "IMMEDIATE" : "READ_READY");
-    // Look for data and if we find any, consume.
+    TSDebug(PLUGIN_NAME.c_str(), "Data event - %s", event == TS_EVENT_IMMEDIATE ? "IMMEDIATE" : "READ_READY");
+    // Look for data and if we find any, consume it.
     if (TSVIOBufferGet(input_vio)) {
       TSIOBufferReader reader = TSVIOReaderGet(input_vio);
-      int64_t n               = TSIOBufferReaderAvail(reader);
+      size_t const n          = TSIOBufferReaderAvail(reader);
       if (n > 0) {
+        char buffer[n];
```

Comment:
Don't bother with the intermediate buffer.
```
auto offset = data->body_bytes.size();
data->body_bytes.resize(offset+n);
TSIOBufferReaderCopy(reader, data->body_bytes.data() + offset, n);
```

Reply:
Looks fine.